cpp/arcae/shared_fcntl_mutex.cc

```diff
@@ -0,0 +1,254 @@
+      case EACCES:
+        fd = -1;
+        break;
+      default:
+        return Status::IOError("Create/Open of ", lock_filename, " failed with  ",
+                               strerror(errno), " (", errno, ')');
+    }
+  }
+
+  return std::make_shared<enable_rwlock>(fd, lock_filename);
+}
+
+SharedFcntlMutex::~SharedFcntlMutex() {
+  if (fd_ != -1) {
+    // Note that this destructor doesn't handle all edge cases.
+    // For example, if another thread holds a read lock, this
+    // code will close the file descriptor, but the other thread
+    // will not know about it. It's assumed that all locks are
+    // released before the destructor is called.
+    close(fd_);
+  }
+  lock_filename_.clear();
+}
+
+void SharedFcntlMutex::unlock() {
+  if (fd_ != -1) {
+    struct flock lock_data = {
+        .l_type = F_UNLCK,
+        .l_whence = SEEK_SET,
+        .l_start = 0,
+        .l_len = 0,
+    };
+
+    fcntl(fd_, F_SETLK, &lock_data);
+  }
+  mutex_.unlock();
+}
+
+void SharedFcntlMutex::unlock_shared() {
+  std::unique_lock<std::mutex> fnctl_lock(fcntl_read_mutex_);
+  // Last reader releases the fcntl lock
+  if (fd_ != -1 && reader_counts_ > 0 && --reader_counts_ == 0) {
+    struct flock lock_data = {
+        .l_type = F_UNLCK,
+        .l_whence = SEEK_SET,
+        .l_start = 0,
+        .l_len = 0,
+    };
+    fcntl(fd_, F_SETLK, &lock_data);
+  };
+
+  mutex_.unlock_shared();
+}
+
+auto SharedFcntlMutex::fcntl_readers() -> std::size_t {
+  std::unique_lock<std::mutex> fnctl_lock(fcntl_read_mutex_);
+  return reader_counts_;
+}
+
+auto SharedFcntlMutex::other_locks(FcntlLockType lock_type)
+    -> Result<SharedFcntlMutex::LockInfo> {
+  struct flock lock_data = {
+      .l_type = lock_type,
+      .l_whence = SEEK_SET,
+      .l_start = 0,
+      .l_len = 0,
+      .l_pid = getpid(),
+  };
+
+  if (fcntl(fd_, F_GETLK, &lock_data) == -1) {
+    return Status::IOError("Fail to retrieve lock information for ", lock_filename_);
+  }
+
+  return SharedFcntlMutex::LockInfo{lock_data.l_type, lock_data.l_pid};
+}
+
+auto SharedFcntlMutex::lock_impl(bool write) -> Status {
+  std::unique_lock<std::mutex> reader_lock(fcntl_read_mutex_, std::defer_lock);
+
+  if (write) {
+    if (!has_fd()) return Status::Cancelled("Readonly Table");
+    mutex_.lock();
+  } else {
+    mutex_.lock_shared();
+    // If there's no fcntl lock we're done
+    if (!has_fd()) return Status::OK();
+    // Acquire the reader lock
+    // If the underlying fcntl lock has already been acquired
+    // then increment again and return successfully.
+    // Otherwise the read lock now guards the acquisition
+    // of the fcntl lock below
+    reader_lock.lock();
+    if (reader_counts_++ > 0) return Status::OK();
+  }
+
+  // Repeatedly attempt acquisition of
+  // the fcntl lock with exponential backoff.
+  // A blocking F_SETLKW could be used but in practice
+  // this can result in the kernel producing EDEADLK.
```

Comment:
I see the man page says it can produce false positives, but have you considered trying it first then falling back to the backoff strategy if you do receive EDEADLK?